### ISSUE TITLE
pkg_resources.NullProvider since pytest 4.6.0+

### ISSUE THREAD

**Issue by tobywf** [pytest-dev/pytest#5392]
When using a pytest version 4.6.0 or up, Jinja's `PackageLoader` breaks with "NotImplementedError: Can't perform this operation for unregistered loader type". This is because `pkg_resources` returns a `NullProvider`.

Minimal working example:

```
python3 -m venv env && source env/bin/activate
pip install 'pytest>=4.6.0' 'jinja2==2.10.1'
mkdir -p tests/templates
touch tests/__init__.py
echo "hello, world" > tests/templates/test.txt
cat << EOF > tests/test_jinja.py
from jinja2 import Environment, PackageLoader

def test_package_loader():
    env = Environment(loader=PackageLoader(__name__))
    assert "hello" in env.get_template("test.txt").render()
EOF
pytest tests/
pip install 'pytest==4.5.0'
pytest tests/
echo "done"
```

With `pytest>=4.6.0` (confirmed with `4.6.0`, `4.6.1`, and `4.6.2`), it fails with:
```
self = <pkg_resources.NullProvider object at 0x10c2d69b0>, path = '/workplace/tobflem/pytest/tests/templates/test.txt'

    def _has(self, path):
        raise NotImplementedError(
>           "Can't perform this operation for unregistered loader type"
        )
E       NotImplementedError: Can't perform this operation for unregistered loader type

env/lib/python3.7/site-packages/pkg_resources/__init__.py:1457: NotImplementedError
```
After pytest is pinned to `4.5.0` again, no failure is observed.

There was an issue way back, #366, which seems to be quite similar, where "`__loader__` is being wrapped in an AssertionRewritingHook". Could be a similar issue?

Hope this is enough info, let me know if you need more!

**Comment by tobywf** [comment]
Pretty sure it's to do with commit 13f02af97d676bdf7143aa1834c8898fbf753d87, where `_register_with_pkg_resources` was removed from `AssertionRewritingHook` in `src/_pytest/assertion/rewrite.py`. When I add that back in, the issue goes away.

**Comment by blueyed** [comment]
> Environment(loader=PackageLoader(__name__))

Do you have to use this?

It seems like you are using a different loader then in your real code already probably, and should probably use an `Environment` from your real app?

**Comment by tobywf** [comment]
That's just a minimal example I initially came up with (before I looked at writing the unit test without Jinja), the production code is somewhat more complicated. We can definitely work around this if required, or stick to `pytest-4.5.0` for now. I don't 100% grok the issue yet myself.

**Comment by blueyed** [comment]
From what I understand there is (just) no PackageLoader anymore.
Sounds like @asottile suggested a good workaround (https://github.com/pytest-dev/pytest/pull/5397#issuecomment-498890646).

As it stands #5397 is not going to fly, but could maybe be turned into one that documents this - although not clear to me if that is really useful / where to put this.

**Comment by tobywf** [comment]
For reference, [on that PR @asottile pointed out](https://github.com/pytest-dev/pytest/pull/5397#issuecomment-498890646) there's a feature request https://github.com/pallets/jinja/issues/970 / PR https://github.com/pallets/jinja/pull/976 for a Jinja loader that uses `importlib` instead of `pkg_resources`. However, it isn't merged yet. This works for me, I can stick with the `4.5.0` release until the new loader lands in Jinja. (This means semver is a bit off though.)

Does the [deprecations page](https://docs.pytest.org/en/latest/deprecations.html) need updating? Otherwise, feel free to close this issue. Thanks for the quick responses!

**Comment by nicoddemus** [comment]
> Does the deprecations page need updating? Otherwise, feel free to close this issue. Thanks for the quick responses!

Thanks @tobywf.

Not sure we should be adding something to the deprecations page, as we didn't actually deprecate any pytest features, what changed was an internal implementation detail (IIUC). Let's see, if more people come up with problems we should add a blurb to the docs somewhere pointing to a workaround.

Thanks again!

**Comment by omry** [comment]
This is unfortunate. can you share more details about why pytest needs to interfere with pkg_resource?

This is breaking the tests of any library using pytest to test code paths using pkg_resource.

**Comment by tobywf** [comment]
Can you implement the fix suggested in https://github.com/pytest-dev/pytest/pull/5397#issuecomment-498890646 (using `importlib.resources` or the backport)?

It's been a while, and I can't speak for the pytest team. But I believe the issue is that pytest in fact *doesn't* interfere with `pkg_resources` any more. And `pkg_resources` has some downsides, which is [why `importlib.resources` was made](https://docs.python.org/3.7/library/importlib.html#module-importlib.resources) (not a pytest-specific thing). So using that is the future. I've found it quite [easy to migrate](https://importlib-resources.readthedocs.io/en/latest/migration.html), and the APIs are more sane. If you can't, then pinning pytest to `4.5.0` also works, but that's just putting it off.

**Comment by omry** [comment]
Hi @tobywf , I went implemented an [importlib-resource based ConfigSource for Hydra](https://github.com/facebookresearch/hydra/commit/0226135e7a9abeb5d9ac82e2efdf18f34364b467) but this is not done as I discovered some functionality gaps in importlib-resources.

I filed [this](https://gitlab.com/python-devs/importlib_resources/-/issues/103) and will see if I can get this functionality added.
This is also a risky change that may cause issues in non standard target environments so it will be while before I can fully migrate to it
in the mean time, I can't really switch to the new ConfigSource (that indeed solves the pytest problem).

Back to pkg_resource:
I can dig in a bit to provide details about why I think pytest is interfering with pkg_resource (I did not spend a whole lot of time debugging this so my analysis may be wrong).

**Comment by omry** [comment]
I would very much like to get the the bottom of this with the pytest team.
I think low level infrastructure like pytest should really strive to be as compatible as possible with downstream projects.

While I would love to switch to imporlib-resources, it does not seem possible in it's current form and I may not be able to do with without some user facing regressions.

**Comment by asottile** [comment]
it *sounds* like you're trying to embed resources inside your testsuite and/or not providing a proper importer for your resources.  If they are outside of your testsuite there'll be no rewriting occurring (and therefore avoid the pytest importer) *or* if you provide resources in a PEP-compatible way they will be available

**Comment by nicoddemus** [comment]
@omry can you provide some sample code that demonstrates the issue you are having? As @tobywf mentions, pytest *almost* doesn't rely on `pkg_resources` anymore (only place where we import it is in [`monkeypatch.syspath_prepend`](https://github.com/pytest-dev/pytest/blob/4e5a7b7d2f1091dbfeff37c8a1d17f65d55beed1/src/_pytest/monkeypatch.py#L298-L318)).

**Comment by omry** [comment]
Thanks. Upon some more digging, it seems like I was missing an `__init__.py` in my resources dir.

There was another factor of which directory the the test was being executed from which masked the `__init__.py` problem (Hydra supports both file based and module based resource loading).

**Comment by omry** [comment]
Okay, last comment on this issue:

Turns out the reason I ran into such difficulties had to do with me creating a pytest plugin.

I created a new [issue with a minimal example](https://github.com/pytest-dev/pytest/issues/7375). It demonstrates that registering a pytest plugin (with setup.py entry points) is enough to break  user code using pkg_resource in that app.

Would be great to get some eyes on it.
